controller/api/public/stat_summary.go

```diff
@@ -0,0 +1,334 @@
+			Ok: &pb.StatSummaryResponse_Ok{
+				StatTables: []*pb.StatTable{
+					&pb.StatTable{
+						Table: &pb.StatTable_PodGroup_{
+							PodGroup: &pb.StatTable_PodGroup{
+								Rows: rows,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return &rsp, nil
+}
+
+func promLabel(key, val string) string {
+	return fmt.Sprintf("%s=\"%s\"", key, val)
+}
+
+func buildRequestLabels(req *pb.StatSummaryRequest) string {
+	labels := []string{}
+
+	var direction string
+	switch req.Outbound.(type) {
+	case *pb.StatSummaryRequest_None:
+		direction = "inbound"
+
+	case *pb.StatSummaryRequest_OutToResource:
+		direction = "outbound"
+
+		dstLabel := fmt.Sprintf("dst_namespace=\"%s\", dst_%s=\"%s\"",
+			req.GetOutToResource().Namespace,
+			req.GetOutToResource().Type,
+			req.GetOutToResource().Name,
+		)
+		labels = append(labels, dstLabel)
+
+	case *pb.StatSummaryRequest_OutFromResource:
+		direction = "outbound"
+
+		srcLabel := fmt.Sprintf("dst_namespace=\"%s\", dst_%s=\"%s\"",
+			req.Resource.Spec.Namespace,
+			req.Resource.Spec.Type,
+			req.Resource.Spec.Name,
+		)
+		labels = append(labels, srcLabel)
+
+		outFromNs := req.GetOutFromResource().Namespace
+		if outFromNs == "" {
+			outFromNs = req.Resource.Spec.Namespace
+		}
+
+		labels = append(labels, promLabel("namespace", outFromNs))
+		if req.Resource.Spec.Name != "" {
+			labels = append(labels, promLabel(req.GetOutFromResource().Type, req.GetOutFromResource().Name))
+		}
+	default:
+		direction = "inbound"
+
+	}
+
+	// it's weird to check this again outside the switch, but including this code
+	// in the other three switch branches is very repetitive
+	if req.GetOutFromResource() == nil {
+		labels = append(labels, promLabel("namespace", req.Resource.Spec.Namespace))
+		if req.Resource.Spec.Name != "" {
+			labels = append(labels, promLabel(req.Resource.Spec.Type, req.Resource.Spec.Name))
+		}
+	}
+	labels = append(labels, promLabel("direction", direction))
+
+	return strings.Join(labels, ",")
+}
+
+func (h *handler) getRequests(ctx context.Context, reqLabels string, groupBy string, timeWindow string) (map[string]*pb.BasicStats, error) {
+	requestsQuery := fmt.Sprintf(reqQuery, reqLabels, timeWindow, groupBy)
+
+	resultVector, err := h.queryProm(ctx, requestsQuery)
+	if err != nil {
+		return nil, err
+	}
+	return processRequests(resultVector, groupBy), nil
+}
+
+func processRequests(vec model.Vector, labelSelector string) map[string]*pb.BasicStats {
+	result := make(map[string]*pb.BasicStats)
+	for _, sample := range vec {
+		labels := metricToMap(sample.Metric)
+		if result[labels[labelSelector]] == nil {
+			result[labels[labelSelector]] = &pb.BasicStats{}
+		}
+
+		switch labels["classification"] {
+		case "success":
+			result[labels[labelSelector]].SuccessCount = uint64(sample.Value)
+		case "fail":
+			result[labels[labelSelector]].FailureCount = uint64(sample.Value)
+		}
+	}
+	return result
+}
+
+func metricToMap(metric model.Metric) map[string]string {
```

Comment:
where we're using `map[string]string` as prometheus labels, consider using a Prometheus [LabelSet](https://godoc.org/github.com/prometheus/common/model#LabelSet)

Reply:
I can switch it. It might save us some extra mapping where there are a lot of labels not to have to run metricToMap on all metrics' labels. The repeated casting to `model.LabelName` and back to string is slightly less readable though.

```
func processRequests(vec model.Vector, labelSelector string) map[string]*pb.BasicStats {
	result := make(map[string]*pb.BasicStats)

	for _, sample := range vec {
		label := string(sample.Metric[model.LabelName(labelSelector)])
		if result[label] == nil {
			result[label] = &pb.BasicStats{}
		}

		switch string(sample.Metric[model.LabelName("classification")]) {
		case "success":
			result[label].SuccessCount = uint64(sample.Value)
		case "fail":
			result[label].FailureCount = uint64(sample.Value)
		}
	}
	return result
}
```

Reply:
Yeah, agree it is slightly less readable this way. Ideally we move more of this part of the codebase to just passing canonical Prometheus data around, so in this case I think the tradeoff is worth it.